api/v1alpha3/zz_generated.conversion.go

```diff
@@ -0,0 +1,1389 @@
+	out.ObjectMeta = in.ObjectMeta
+	if err := Convert_v1alpha3_AWSClusterSpec_To_v1alpha4_AWSClusterSpec(&in.Spec, &out.Spec, s); err != nil {
+		return err
+	}
+	if err := Convert_v1alpha3_AWSClusterStatus_To_v1alpha4_AWSClusterStatus(&in.Status, &out.Status, s); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Convert_v1alpha3_AWSCluster_To_v1alpha4_AWSCluster is an autogenerated conversion function.
+func Convert_v1alpha3_AWSCluster_To_v1alpha4_AWSCluster(in *AWSCluster, out *v1alpha4.AWSCluster, s conversion.Scope) error {
+	return autoConvert_v1alpha3_AWSCluster_To_v1alpha4_AWSCluster(in, out, s)
+}
+
+func autoConvert_v1alpha4_AWSCluster_To_v1alpha3_AWSCluster(in *v1alpha4.AWSCluster, out *AWSCluster, s conversion.Scope) error {
+	out.ObjectMeta = in.ObjectMeta
+	if err := Convert_v1alpha4_AWSClusterSpec_To_v1alpha3_AWSClusterSpec(&in.Spec, &out.Spec, s); err != nil {
+		return err
+	}
+	if err := Convert_v1alpha4_AWSClusterStatus_To_v1alpha3_AWSClusterStatus(&in.Status, &out.Status, s); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Convert_v1alpha4_AWSCluster_To_v1alpha3_AWSCluster is an autogenerated conversion function.
+func Convert_v1alpha4_AWSCluster_To_v1alpha3_AWSCluster(in *v1alpha4.AWSCluster, out *AWSCluster, s conversion.Scope) error {
+	return autoConvert_v1alpha4_AWSCluster_To_v1alpha3_AWSCluster(in, out, s)
+}
+
+func autoConvert_v1alpha3_AWSClusterList_To_v1alpha4_AWSClusterList(in *AWSClusterList, out *v1alpha4.AWSClusterList, s conversion.Scope) error {
+	out.ListMeta = in.ListMeta
+	out.Items = *(*[]v1alpha4.AWSCluster)(unsafe.Pointer(&in.Items))
+	return nil
+}
+
+// Convert_v1alpha3_AWSClusterList_To_v1alpha4_AWSClusterList is an autogenerated conversion function.
+func Convert_v1alpha3_AWSClusterList_To_v1alpha4_AWSClusterList(in *AWSClusterList, out *v1alpha4.AWSClusterList, s conversion.Scope) error {
+	return autoConvert_v1alpha3_AWSClusterList_To_v1alpha4_AWSClusterList(in, out, s)
+}
+
+func autoConvert_v1alpha4_AWSClusterList_To_v1alpha3_AWSClusterList(in *v1alpha4.AWSClusterList, out *AWSClusterList, s conversion.Scope) error {
+	out.ListMeta = in.ListMeta
+	out.Items = *(*[]AWSCluster)(unsafe.Pointer(&in.Items))
+	return nil
+}
+
+// Convert_v1alpha4_AWSClusterList_To_v1alpha3_AWSClusterList is an autogenerated conversion function.
+func Convert_v1alpha4_AWSClusterList_To_v1alpha3_AWSClusterList(in *v1alpha4.AWSClusterList, out *AWSClusterList, s conversion.Scope) error {
+	return autoConvert_v1alpha4_AWSClusterList_To_v1alpha3_AWSClusterList(in, out, s)
+}
+
+func autoConvert_v1alpha3_AWSClusterSpec_To_v1alpha4_AWSClusterSpec(in *AWSClusterSpec, out *v1alpha4.AWSClusterSpec, s conversion.Scope) error {
+	if err := Convert_v1alpha3_NetworkSpec_To_v1alpha4_NetworkSpec(&in.NetworkSpec, &out.NetworkSpec, s); err != nil {
+		return err
+	}
+	out.Region = in.Region
+	out.SSHKeyName = (*string)(unsafe.Pointer(in.SSHKeyName))
+	// TODO: Inefficient conversion - can we improve it?
```

Comment:
In case you're not already aware anyway. CAPZ solved it like this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/55c757c661f689fff8fb818e7171472ad37122a1/api/v1alpha3/azurecluster_conversion.go#L221

Reply:
If this is about compileErrorOnMissingConversion() error during generation conversion, it is solved by changing pointer slices to slices.
https://github.com/kubernetes-sigs/cluster-api/issues/4096#issuecomment-765597734

Or is this another issue?

Reply:
I think it's about providing conversion functions for types which are not in the local go module.
I.e. in CAPZ they provided Convert functions for APIEndpoint (APIEndpoint is from the cluster-api) repo in the same package and conversion-gen then just used them in the generated code.

Afaik APIEndpoint is not used in slices (https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/api/v1alpha3/awscluster_types.go#L44) so it should be another issue.

Just stumpled upon the issue when implementing CAPO v1alpha4 and saw the TODO here and the (probably) correct implementation in CAPZ and though I leave you a hint here.